### PR TITLE
SNOW-1348656 Send records that could not be parsed to DLQ

### PR DESCRIPTION
**Pull request by sfc-gh-mbobowski**
# Overview

SNOW-1348656 Kafka Processing error not sent to DLQ correctly.

Prior to this change Snowpipe Streaming with schematization did not send records that could not be parsed to DLQ but throw exception and stop the ingestion instead.

Instead of catching all `SnowflakeKafkaConnectorException` (as suggested in the JIRA) I went with more conservative approach and to it for specific error code only.

## Pre-review checklist
- [ ] This change should be part of a Behavior Change Release. See [go/behavior-change](http://go/behavior-change).
- [ ] This change has passed Merge gate tests
- [ ] Snowpipe Changes
- [x] Snowpipe Streaming Changes
- [ ] This change is TEST-ONLY
- [ ] This change is README/Javadocs only
- [ ] This change is protected by a config parameter <PARAMETER_NAME> eg `snowflake.ingestion.method`.
    - [ ] `Yes` - Added end to end and Unit Tests. 
    - [ ] `No` - Suggest why it is not param protected
- [ ] Is his change protected by parameter <PARAMETER_NAME> on the server side?
    - [ ] The parameter/feature is not yet active in production (partial rollout or PrPr, see [Changes for Unreleased Features and Fixes](http://go/ppp-prpr)).
    - [ ] If there is an issue, it can be safely mitigated by turning the parameter off. This is also verified by a test (See [go/ppp](http://go/ppp)).

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 1]
```diff
@@ -1228,7 +1230,7 @@ protected long getApproxSizeOfRecordInBytes(SinkRecord kafkaSinkRecord) {
           }
         }
       }
-    } catch (JsonProcessingException e) {
+    } catch (Exception e) {
       // We ignore any errors here because this is just calculating the record size
```
Comment (sfc-gh-wtrefon) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1234]: could we at least log sth here?

File: src/test/java/com/snowflake/kafka/connector/SnowflakeSinkTaskStreamingDLQTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import net.snowflake.ingest.utils.SFException;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mockito;
+
+public class SnowflakeSinkTaskStreamingDLQTest {
```
Comment (sfc-gh-wtrefon) [inline src/test/java/com/snowflake/kafka/connector/SnowflakeSinkTaskStreamingDLQTest.java:23]: i would add one more test case:
when we throw exception that has different error code and is not sent to DLQ but rethrown

**Review by sfc-gh-wtrefon — APPROVED** [review]
just one test case missing, overall LGTM

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 3]
```diff
@@ -1320,6 +1322,16 @@ public Pair<List<Map<String, Object>>, List<Long>> getData() {
                 kafkaSinkRecord.kafkaOffset(),
                 kafkaSinkRecord.topic());
             kafkaRecordErrorReporter.reportError(kafkaSinkRecord, e);
```
Comment (sfc-gh-gjachimko) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1324]: shouldn't also records, which raised JsonProcessingException end up in the DLQ? 
i mean - we' ve consumed exception ,reported an error and the message is gone.

Comment (sfc-gh-wtrefon) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1324]: but they end up, we call kafkaRecordErrorReporter.reportError(kafkaSinkRecord, e);
What do u mean?

Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1324]: +1 they are in DLQ.

**Review by sfc-gh-gjachimko — APPROVED** [review]

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 1 above)
Comment (sfc-gh-lshcharbaty) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1234]: Let's catch `SnowflakeKafkaConnectorException` with `ERROR_0010` code only. similar to what you do in `getData()` below.

**Review by sfc-gh-lshcharbaty — APPROVED** [review]

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 1 above)
Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java]: Good point. Done.

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 4]
```diff
@@ -1228,8 +1230,16 @@ protected long getApproxSizeOfRecordInBytes(SinkRecord kafkaSinkRecord) {
...
-      // We ignore any errors here because this is just calculating the record size
+    } catch (Exception e) {
+      boolean isJsonProcessingEx = e instanceof JsonProcessingException;
+      boolean isSnowflakeParsingEx =
+          e instanceof SnowflakeKafkaConnectorException
+              && ((SnowflakeKafkaConnectorException) e).checkErrorCode(SnowflakeErrors.ERROR_0010);
+      if (!(isJsonProcessingEx || isSnowflakeParsingEx)) {
+        // We ignore any errors parsing exceptions here because this is just calculating the record
+        // size
+        throw new RuntimeException(e);
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1241]: Why we're not throwing the original exception?

**Review by sfc-gh-tzhang — APPROVED** [review]
lgtm, thanks!

**Review by sfc-gh-japatel — APPROVED** [review]
can you verify if data gets into DLQ in an end to end test scenario?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 4 above)
Comment (sfc-gh-mbobowski) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1241]: `JsonProcessingException` is checked exception so the compiler requires to add throws clause to the signature of `getApproxSizeOfRecordInBytes`. The call stack leads to `SnowflakeSinkTask#put()` which is Kafka Connect method that we cannot modify. At some point we have to wrap checked exception into the unchecked one.

File: src/test/java/com/snowflake/kafka/connector/SnowflakeSinkTaskStreamingDLQTest.java (hunk 2 above)
Comment (sfc-gh-mbobowski) [reply src/test/java/com/snowflake/kafka/connector/SnowflakeSinkTaskStreamingDLQTest.java:23]: There is no obvious code path to provide such test at task level (checked the one with null records but it turned out to be dead code).
It is possible to mock RecordService to achieve such behaviour but it requires a lot of additional refactoring. I think it is an overkill for that particular fix.

**Pull request merged**